src/platform_impl/linux/wayland/seat/keyboard/handlers.rs

```diff
@@ -149,3 +142,594 @@ pub(super) fn handle_keyboard(
+/// `EventLoopHandle::insert_source()` will be ignored. Instead it triggers the
+/// callback you provided to [`map_keyboard`](fn.map_keyboard.html).
+pub struct RepeatSource {
+    timer: calloop::timer::Timer<()>,
+    state: Rc<RefCell<KbState>>,
+    current_repeat: Rc<RefCell<Option<RepeatData>>>,
+}
+
+impl calloop::EventSource for RepeatSource {
+    type Event = Event<'static>;
+    type Metadata = wl_keyboard::WlKeyboard;
+    type Ret = ();
+
+    fn process_events<F>(
+        &mut self,
+        readiness: calloop::Readiness,
+        token: calloop::Token,
+        mut callback: F,
+    ) -> std::io::Result<calloop::PostAction>
+    where
+        F: FnMut(Event<'static>, &mut wl_keyboard::WlKeyboard),
+    {
+        let current_repeat = &self.current_repeat;
+        let state = &self.state;
+        self.timer
+            .process_events(readiness, token, |(), timer_handle| {
+                if let Some(ref mut data) = *current_repeat.borrow_mut() {
+                    // there is something to repeat
+                    let mut state = state.borrow_mut();
+                    let mut ker = state.process_key_repeat_event(data.keycode + 8);
+
+                    let physical_key = ker.keycode();
+                    let (logical_key, location) = ker.key();
+                    let text = ker.text();
+                    let (key_without_modifiers, _) = ker.key_without_modifiers();
+                    let text_with_all_modifiers = ker.text_with_all_modifiers();
+
+                    let new_time = data.gap + data.time;
+                    // Notify the callback.
+                    callback(
+                        Event::Repeat {
+                            time: (new_time / 1000) as u32,
+                            physical_key,
+                            logical_key,
+                            text,
+                            location,
+                            key_without_modifiers,
+                            text_with_all_modifiers,
+                        },
+                        &mut data.keyboard,
+                    );
+                    // Update the time of last event.
+                    data.time = new_time;
+                    // Schedule the next timeout.
+                    timer_handle.add_timeout(Duration::from_micros(data.gap), ());
+                }
+            })
+    }
+
+    fn register(
+        &mut self,
+        poll: &mut calloop::Poll,
+        token_factory: &mut calloop::TokenFactory,
+    ) -> std::io::Result<()> {
+        self.timer.register(poll, token_factory)
+    }
+
+    fn reregister(
+        &mut self,
+        poll: &mut calloop::Poll,
+        token_factory: &mut calloop::TokenFactory,
+    ) -> std::io::Result<()> {
+        self.timer.reregister(poll, token_factory)
+    }
+
+    fn unregister(&mut self, poll: &mut calloop::Poll) -> std::io::Result<()> {
+        self.timer.unregister(poll)
+    }
```

Comment:
Could you recheck latest SCTK release wrt that code, it was changed iirc given that calloop updated, but maybe that's a new version of it?

Reply:
This is all based off of a 5 months old version of Winit, so maybe that'll be an issue once we rebase.